src/Microsoft.OData.Core/ODataParameterWriterCore.cs

```diff
@@ -816,5 +879,135 @@ private void ValidateTransition(ParameterWriterState newState)
                     throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataParameterWriterCore_ValidateTransition_UnreachableCodePath));
             }
         }
+
+        /// <summary>
+        /// Asynchronously catch any exception thrown by the action passed in; in the exception case move the writer into
+        /// state ExceptionThrown and then rethrow the exception.
+        /// </summary>
+        /// <param name="action">The delegate to execute asynchronously.</param>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        private async Task InterceptExceptionAsync(Func<Task> action)
+        {
+            try
+            {
+                await action().ConfigureAwait(false);
+            }
+            catch
+            {
+                this.EnterErrorScope();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously catch any exception thrown by the action passed in; in the exception case move the writer into
+        /// state ExceptionThrown and then rethrow the exception.
+        /// </summary>
+        /// <param name="func">The delegate to execute asynchronously.</param>
+        /// <returns>A task that represents the asynchronous operation. 
+        /// The value of the TResult parameter contains a T instance.</returns>
+        private async Task<T> InterceptExceptionAsync<T>(Func<Task<T>> func)
+        {
+            try
+            {
+                return await func().ConfigureAwait(false);
+            }
+            catch
+            {
+                this.EnterErrorScope();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously start writing a parameter payload - implementation of the actual functionality.
+        /// </summary>
+		/// <returns>A task that represents the asynchronous write operation.</returns>
+        private async Task WriteStartImplementationAsync()
+        {
+            Debug.Assert(this.State == ParameterWriterState.Start, "this.State == ParameterWriterState.Start");
+
+            await this.InterceptExceptionAsync(this.StartPayloadAsync)
+                .ConfigureAwait(false);

```

Comment:
one line code

Reply:
Below this line we also have:
```csharp
this.EnterScope(ParameterWriterState.CanWriteParameter);
```
We `await` the call to `this.InterceptExceptionAsync` since we need the `this.EnterScope` to be executed in the continuation.